File: Main/ArchiMeter.Analysis/RequirementToTestReport.cs
namespace ArchiMeter.Analysis
{
	using System;
	using System.Collections.Generic;
	using System.Linq;

	public class RequirementToTestReport
	{
		public RequirementToTestReport(int requirementId, IEnumerable<TestData> coveringTests)
		{
			var tests = coveringTests.ToArray();
			this.RequirementId = requirementId;
			this.CoveringTestNames = tests.Select(d => d.TestName).ToArray();
			this.CoveringTests = tests.Select(d => d.TestCode).ToArray();
			this.AssertsPerTest = tests.GroupBy(d => d.AssertCount).Select(g => new Tuple<int, int>(g.Key, g.Count())).ToArray();
		}

		public IEnumerable<Tuple<int, int>> AssertsPerTest { get; private set; }

		public IEnumerable<string> CoveringTests { get; private set; }

		public IEnumerable<string> CoveringTestNames { get; private set; }

		public int RequirementId { get; private set; }
	}
}

File: ArchiMeter.Analysis.FS/RequirementToTestReport.fs
namespace ArchiMeter.Analysis
open System

type RequirementToTestReport(requirementId : int, coveringTests : seq<TestData>) =
    member this.RequirementId = requirementId
    member this.CoveringTestNames = coveringTests |> Seq.map (fun d -> d.TestName) |> Seq.toArray
    member this.CoveringTests = coveringTests |> Seq.map(fun d -> d.TestCode) |> Seq.toArray
    member this.AssertsPerTest = coveringTests |> Seq.groupBy(fun d -> d.AssertCount) |> Seq.map (fun (k, g) -> (k, Seq.length g)) |> Seq.toArray